zcash_client_backend/src/welding_rig.rs

```diff
@@ -81,30 +79,81 @@ fn scan_output<P: consensus::Parameters>(
             to,
             is_change,
             witness,
-            nf: Some(nf),
+            nf,
         });
     }
+
     None
 }
 
-/// Scans a [`CompactBlock`] with a set of [`ViewingKey`]s.
+pub trait ScanningKey {
+    type Nf;
+
+    fn try_decryption<P: consensus::Parameters>(
+        &self,
+        params: &P,
+        height: BlockHeight,
+        epk: &jubjub::ExtendedPoint,
+        cmu: &bls12_381::Scalar,
+        ct: &[u8],
+    ) -> Option<(Note, PaymentAddress)>;
+
+    fn nf(&self, note: &Note, witness: &IncrementalWitness<Node>) -> Self::Nf;
+}
+
+impl ScanningKey for ViewingKey {
```

Comment:
This should be implemented for `ExtendedFullViewingKey` (being the standard type we end up passing around for Sapling), as was originally the case pre-DAA. `ViewingKey` is a type we shouldn't be exposing, but haven't had the time to refactor away yet. `FullViewingKey` is annoyingly also basically internal at this point for Sapling due to the way we did ZIP 32 derivation for Sapling. This won't be the case for Orchard, where there isn't an `extfvk` type.